### PR TITLE
fixed an issue with the check of the errorField code. Now the error i…

### PR DESCRIPTION
**Pull request by loretrisolini**
…s not raised if the errorField code is equal to 200

### CODE REVIEW

File: lib/services/ngsiService.js [hunk 1]
```diff
@@ -129,7 +129,9 @@ function sendRegistrations(unregister, deviceData, callback) {
         } else if (response && body && response.statusCode === 200) {
             var errorField = getErrorField(body);
 
-            if (errorField) {
+			/* fixed an issue with the check of the errorField code. Now the error is not raised if the errorField code is equal to 200 (ReasonPhrase: OK) */
```
Comment (dmoranj) [inline lib/services/ngsiService.js]: Please don't leave this comments here. Fix explanations belong to the CHANGES_NEXT_RELEASE file + Github issues.

File: lib/services/ngsiService.js [hunk 2]
```diff
@@ -129,7 +129,9 @@ function sendRegistrations(unregister, deviceData, callback) {
...
+
+            if (errorField && errorField.code!=200) {
```
Comment (dmoranj) [inline lib/services/ngsiService.js]: Better add a check for the existence of `errorField.code`, as the format for the error is quite free and unpredictable.

**Comment by dmoranj** [comment]
Mmmm. I'm afraid the PR has a major problem: PRs must always be merged to develop, not to master:  merges into master are reserved for releases. Please, update your `master` branch with the official `develop` branch, and recreate the PR using develop branches. I'm afraid this part of the contribution policy may not be clear in the Docs, so I'll try to update it ASAP, so this problem won't rise in the future.

**Comment by loretrisolini** [comment]
Hi Daniel,
can this prototype of the iot-agent accepting COAP (
https://github.com/telefonicaid/lightweightm2m-iotagent) communicate
without any problem with the latest version of the iot-discovery?

Lorenzo

2015-06-30 17:13 GMT+02:00 Daniel Moran <notifications@github.com>:

**Comment by dmoranj** [comment]
In theory, they should, as both are using NGSI to communicate. In practice, I haven't tested that particular interaction. If you have, and find any problem, you can file a bug in either one of the components and we will try to solve it.

**Comment by loretrisolini** [comment]
Hi Daniel,

Sorry for the correction that i have done yesterday because i revised it
and it is not correct. In my opinion, the question is more complicated
because the problem that i found is the following:

the ContextBroker, in case of success in the operation performer or
requested, uses this schema:
....
"statusCode": {
     "code": "200",
     "reasonPhrase": "OK"
}
.....

instead the latest version of the iot-discovery response in case of
successful operation is:

.....
"errorCode": {
     "code": "200",
     "reasonPhrase": "OK",
     details: "Stored"
}
.....
So if the iot-agent communicates with the lastest version of the
iot-discovery, the iot-agent gives always an error although the
registerContext operation, in the iot-discovery, is successful.

Best regards,
Lorenzo Trisolini

2015-07-01 10:01 GMT+02:00 Daniel Moran <notifications@github.com>:

**Comment by dmoranj** [comment]
Hi Lorenzo, 

I think I have to check this with the Data Chapter people, to make sure what of the two different approaches is correct. Please, stay tuned.

**Comment by dmoranj** [comment]
I've spoken with the architect of Fiware's IoT Chapter, regarding the integration of IoT Discovery and IoT Agents, and it seems the integration of both components, officially is in the roadmap, but its still experimental. An integration in the public instance is planned after summer. Until then, you are a pioneer :)

Regarding the different way of treating status and error codes, I'm waiting for a colleague in the Data Chapter that is currently on holidays to return this monday. I will check with him and give you a definite answer about this.

**Comment by loretrisolini** [comment]
Thank you for your time,

Lorenzo Trisolini

2015-07-03 10:32 GMT+02:00 Daniel Moran <notifications@github.com>:

**Comment by loretrisolini** [comment]
Hi Daniel,
Does the iot-agent support the subscribeContextRequest operation (coming
from the iotBroker GE)?

Lorenzo Trisolini

2015-07-07 9:52 GMT+02:00 Lorenzo Trisolini <lore.trisolini@gmail.com>:

**Comment by dmoranj** [comment]
Not as such, but you may overcome that issue using the Context Broker. The IoT Agents were designed to be attached to a Context Broker (typically an instance of  [Orion](https://github.com/telefonicaid/fiware-orion)) that would act as the main point of interaction between all the components of your application. If you use it in that way, you can always subscribe to the notifications in the CB instead of doing it in the IOTA. That being said, accepting subscriptions may be a useful feature, so you can add a feature and we will discuss its inclusion in the project (we are also open to contributions :) ).

**Comment by fgalan** [comment]
Sorry for joining late to the discussion...

Regarding:

> the ContextBroker, in case of success in the operation performer or
> requested, uses this schema:
>
> "statusCode": {
>      "code": "200",
>     "reasonPhrase": "OK"
>}
>
> instead the latest version of the iot-discovery response in case of
> successful operation is:
>
> "errorCode": {
>     "code": "200",
>     "reasonPhrase": "OK",
>     details: "Stored"
> }

I would need to see the complete response payload where these fragments appear in order to provide feedback... could you add that information in a new comment, please?

**Comment by loretrisolini** [comment]
Hi Daniel,
Thank you for your time and your response. I will keep you update for the
adding of the subscriptions feature to the IoTAgent.

Best Regards,
Lorenzo Trisolini

2015-07-07 12:00 GMT+02:00 Daniel Moran <notifications@github.com>:

**Comment by loretrisolini** [comment]
Hi,
sorry for my first mail because it was not clear at all, I revised the
problem that I had. I connect the IoT Agent to the IoT Discovery. When the
IoT Agent sends the registerContext() to the IoTDIscovery, the latter
answers with this message:

{
"duration":"P1M",
"registrationId":"UniS_l3EAFE5MDP",
*"errorCode":{*
*       "code":200,*
*       "reasonPhrase":"OK",*
*       "details":"Stored"*
*}*
}

instead the ContextBroker, for the same operation, answers with this
message:

{
  "duration" : "P1M",
  "registrationId" : "52a744b011f5816465943d58"
}

So the function sendRegistrations() in the file
$MYHOME/lightweightm2m-iotagent/node_modules/iotagent-node-lib/lib/services/ngsiService.js,
returns always a registration error, when the IoTAgent send
registerContext() to the IoTDiscovery.
I think the problem it is in the function getErrorField() (called in
sendRegistration()), because it is made to analyze the response received by
a ContextBroker that is different from the IoTDiscovery response (I tested
the difference only for the registerContext() from the IoTAgent to the
IoTDiscovery).
The control, in sendRegistration(), "if (errorField)" (after "var
errorField = getErrorField(body)") will be always positive because the
IoTDiscovery returns an errorCode field in case of successful
registerContext().
I think that this question arises because The IoT Agents were designed to
be attached to a Context Broker as Daniel said to me.

Best Regards,
Lorenzo Trisolini


2015-07-08 13:31 GMT+02:00 Fermín Galán Márquez <notifications@github.com>:

**Comment by dmoranj** [comment]
This PR has been stalled for quite a long time, and is currently not mergeable. Once it is mergeable with develop, please reopen an we will reevaluate. Thanks.

**Pull request closed without merging**